src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs

```diff
@@ -446,24 +446,39 @@ public override void Abort()
             Dispose();
         }
 
-        public override async Task CloseAsync(WebSocketCloseStatus closeStatus, string? statusDescription, CancellationToken cancellationToken)
+        public override Task CloseAsync(WebSocketCloseStatus closeStatus, string? statusDescription, CancellationToken cancellationToken)
         {
             _writeBuffer = null;
             ThrowIfNotConnected();
-            await CloseAsyncCore(closeStatus, statusDescription, cancellationToken).ConfigureAwait(continueOnCapturedContext: true);
-        }
-
-        private async Task CloseAsyncCore(WebSocketCloseStatus closeStatus, string? statusDescription, CancellationToken cancellationToken)
-        {
-            ThrowOnInvalidState(State, WebSocketState.Open, WebSocketState.CloseReceived, WebSocketState.CloseSent);
 
             WebSocketValidate.ValidateCloseStatus(closeStatus, statusDescription);
 
-            _tcsClose = new TaskCompletionSource();
-            _innerWebSocketCloseStatus = closeStatus;
-            _innerWebSocketCloseStatusDescription = statusDescription;
-            _innerWebSocket!.Invoke("close", (int)closeStatus, statusDescription);
-            await _tcsClose.Task.ConfigureAwait(continueOnCapturedContext: true);
+            try
+            {
+                ThrowOnInvalidState(State, WebSocketState.Open, WebSocketState.CloseReceived, WebSocketState.CloseSent);
+            }
+            catch (Exception exc)
+            {
+                return Task.FromException(exc);
+            }
+
+            return CloseAsyncCore(closeStatus, statusDescription, cancellationToken);
+        }
+
+        private Task CloseAsyncCore(WebSocketCloseStatus closeStatus, string? statusDescription, CancellationToken cancellationToken)
+        {
+            try
+            {
+                _tcsClose = new TaskCompletionSource();
+                _innerWebSocketCloseStatus = closeStatus;
+                _innerWebSocketCloseStatusDescription = statusDescription;
+                _innerWebSocket!.Invoke("close", (int)closeStatus, statusDescription);
+                return _tcsClose.Task;
+            }
+            catch (Exception exc)
+            {
+                return Task.FromException(exc);
+            }
         }
 
         public override Task CloseOutputAsync(WebSocketCloseStatus closeStatus, string? statusDescription, CancellationToken cancellationToken) => throw new PlatformNotSupportedException();
```

Comment:
Unrelated to this PR. Are there plans to implement this method? If not, can it be marked as unsupported in the browser?

I know the browser doesn't provide an API for partial closes like this, but it would be very useful to fake it by clearing the websockets message handlers and calling `onclose` inline.

You can see what we did in SignalR to not have possible long delays between calling `close` and the `onclose` callback being invoked.
https://github.com/dotnet/aspnetcore/blob/0ace31c91138d258008f9547478d156a97f1a87e/src/SignalR/clients/ts/signalr/src/WebSocketTransport.ts#L144-L146
https://github.com/dotnet/aspnetcore/blob/0ace31c91138d258008f9547478d156a97f1a87e/src/SignalR/clients/ts/signalr/src/WebSocketTransport.ts#L156-L170

Reply:
There is an issue set up here to look into this:  https://github.com/dotnet/runtime/issues/45468

Could you please comment on this issue so that we can follow up to get this working for you.